src/System.Security.Permissions/src/System/Security/Permissions/PrincipalPermission.cs

```diff
@@ -2,25 +2,300 @@
+        {
+            _idArray = array;
+        }
+
+        private bool IsEmpty()
+        {
+            foreach (IDRole idRole in _idArray)
+            {
+                if (idRole.ID == null || idRole.ID.Length != 0 || idRole.Role == null || idRole.Role.Length != 0 || idRole.Authenticated)
+                    return false;
+            }
+            return true;
+        }
+
+        private bool VerifyType(IPermission perm)
+        {
+            // if perm is null, then obviously not of the same type
+            return (perm != null) && (perm.GetType() == GetType());
+        }
+
+        public bool IsUnrestricted()
+        {
+            foreach (IDRole idRole in _idArray)
+            {
+                if (idRole.ID != null || idRole.Role != null || !idRole.Authenticated)
+                    return false;
+            }
+            return true;
+        }
+
+        public bool IsSubsetOf(IPermission target)
+        {
+            if (target == null)
+            {
+                return IsEmpty();
+            }
+            else if (!VerifyType(target))
+            {
+                throw new ArgumentException(SR.Argument_WrongType, GetType().FullName);
+            }
+
+            PrincipalPermission operand = (PrincipalPermission)target;
+
+            if (operand.IsUnrestricted())
+            {
+                return true;
+            }
+            else if (IsUnrestricted())
+            {
+                return false;
+            }
+
+            foreach (IDRole idRole in _idArray)
+            {
+                bool foundMatch = false;
+                foreach (IDRole operandIdRole in operand._idArray)
+                {
+                    if ((operandIdRole.Authenticated == idRole.Authenticated) &&
+                        (operandIdRole.ID == null || (idRole.ID != null && idRole.ID.Equals(operandIdRole.ID))) &&
+                        (operandIdRole.Role == null || (idRole.Role != null && idRole.Role.Equals(operandIdRole.Role))))
+                    {
+                        foundMatch = true;
+                        break;
+                    }
+                }
+
+                if (!foundMatch)
+                    return false;
+            }
+
+            return true;
+        }
+
+        public IPermission Intersect(IPermission target)
+        {
+            if (target == null)
+            {
+                return null;
+            }
+            else if (!VerifyType(target))
+            {
+                throw new ArgumentException(SR.Argument_WrongType, GetType().FullName);
+            }
+            else if (IsUnrestricted())
+            {
+                return target.Copy();
+            }
+
+            PrincipalPermission operand = (PrincipalPermission)target;
+
+            if (operand.IsUnrestricted())
+            {
+                return Copy();
+            }
+
+            List<IDRole> idroles = new List<IDRole>();
```

Comment:
Nit: Before the list was lazily allocated only when needed, now it's always being allocated.